### PR TITLE
Add outlines.generate.fsm() API entrypoint

### PR DESCRIPTION
**Pull request by miftahmoha**
This adds the outlines.generate.fsm() API entrypoint following https://github.com/outlines-dev/outlines/issues/670.

#### Tests:
tests

### CODE REVIEW

File: outlines/fsm/fsm.py [hunk 1]
```diff
@@ -182,6 +182,47 @@ def next_state(self, state: FSMState, token_id: int) -> FSMState:
...
         return FSMState(next_state)
 
+    @classmethod
+    def from_interegular_fsm(
+        cls, interegular_fsm: interegular.fsm.FSM, tokenizer: "Tokenizer"
+    ):
+        from_interegular_instance = cls.__new__(cls)
+
+        @cache()
+        def create_states_mapping_from_interegular_fsm(
```
Comment (lapp0) [inline outlines/fsm/fsm.py:192]: As discussed here, `@cache` doesn't work with FSMs. 

https://github.com/outlines-dev/outlines/issues/670#issuecomment-1958266151

We should leave this uncached for now. There needs to be an upstream fix in `interegular` which makes `AnythingElse` have a consistent hash before we can cache.

We should leave this uncached for now.

**Review by lapp0 — COMMENTED** [review]
Could you please review the test cases `tests/fsm/test_regex.py` and introduce test cases for this new method?

It's probably sufficient to simply verify that `RegexFSM.from_interegular_fsm(interegular.parse_pattern(pattern)) == RegexFSM(pattern)` for a few `pattern`s.

**Comment by miftahmoha** [comment]
@rlouf Done. 

According to https://github.com/outlines-dev/outlines/issues/670#issuecomment-1949402672, `interegular.FSM` offers operations that `RegexFSM` doesn't seem to have. For the examples, we've got https://github.com/outlines-dev/outlines/issues/666 and https://github.com/outlines-dev/outlines/issues/156 that we can use.

What do you think @lapp0?

**Comment by lapp0** [comment]
@miftahmoha yes, we should document those use cases. This perfectly solves those problems. 

Use a pattern but disallow certain keywords.

Example: A list of strings, but none of those strings can be "pink" or "elephant"
```
import interegular

list_of_strings_pattern = """\["[^"\s]*"(?:,"[^"\s]*")*\]"""
pink_elephant_pattern = """.*(pink|elephant).*"""

list_of_strings_fsm = interegular.parse_pattern(list_of_strings_pattern).to_fsm()
pink_elephant_fsm = interegular.parse_pattern(pink_elephant_pattern).to_fsm()



interegular.parse_pattern("
```

**Comment by miftahmoha** [comment]
@lapp0 @rlouf I added some documentation.

**Comment by rlouf** [comment]
Thank you so much for the valuable addition @miftahmoha !

**Pull request merged**